src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -407,6 +407,8 @@ inferBinder val (NamedBinder name binder) = do
   return $ M.insert name val m
 inferBinder val (PositionedBinder pos _ binder) =
   warnAndRethrowWithPosition pos $ inferBinder val binder
+-- | NOTE: Fail!
+inferBinder val (TypedBinder ty binder) = val =?= ty >> inferBinder val binder
```

Comment:
For now, it's restricted to monotypes like you mentioned (mainly because I'm not entirely sure where to start with subsumption checking)